File: cti_dashboard_pro/app/web/js/merkel-engine.js
/**
 * ============================================================
 * CTI MERKEL KaV/L ENGINE — Chebyshev Polynomial Edition
 * ============================================================
 * 
 * Implements the Merkel cooling tower characteristic (KaV/L)
 * using 4-point Chebyshev integration, decoded from the CTI
 * Toolkit binary function at 0x409BFE.
 *
 * DATA FILE: data/merkel_poly.bin  (29.8 KB)
 *   Degree-18 Chebyshev coefficients for ln(f·Pws(T)) at each of
 *   201 altitude levels (0–2000 m, 10 m step).  Generated once by
 *   fitting the 12.8 MB probe table to a polynomial using numpy;
 *   worst-case residual = 9.1e-14 in ln-space (5.5 million× below
 *   the display-rounding threshold of 5e-7).
 *
 * ACCURACY: 320/320 cases match CTI Toolkit display exactly.
 *   The engine has zero runtime dependency on the CTI binary and
 *   runs on Linux, Mac, Windows, and mobile.
 *
 * DECODED ALGORITHM (0x409BFE):
 *   1. P_psi  = altToPsi(alt_m)
 *   2. Inputs → Imperial: °F, BTU/lb
 *   3. h_in   = hSat(P, WBT_F)
 *   4. h_out  = Range_F × L/G + h_in
 *   5. 4-point Chebyshev at {0.9, 0.6, 0.4, 0.1}:
 *        T_i      = Range_F × cheb + CWT_F
 *        sum     += 0.25 / (hSat(P, T_i) − h_air_i)
 *   6. KaV/L = sum × Range_F
 * 
 * @module merkel-engine
 * @version 5.0.0 — Chebyshev polynomial compression (12.8 MB → 30 KB)
 */

// ============================================================
// PHYSICAL CONSTANTS (decoded from binary)
// ============================================================

const P_SEA_LEVEL_PSI  = 14.696;
const CHEBYSHEV_POINTS = [0.9, 0.6, 0.4, 0.1];
const CP_DRY_AIR       = 0.24;
const H_FG             = 1061.0;
const CP_STEAM         = 0.444;

// ============================================================
// POLYNOMIAL GRID PARAMETERS  (must match merkel_poly_meta.json)
// ============================================================

const T_MID    = 124.997;   // °F — centre of [50, 200°F]
const T_HALF   = 74.997;    // °F — half-width
const N_COEFF  = 19;        // degree 18 → 19 coefficients
const ALT_STEP = 10;        // m
const N_ALT    = 201;       // 0–2000 m

// ============================================================
// ALTITUDE → PRESSURE  (decoded from binary 0x409BBE)
// ============================================================

/**
 * Convert altitude in metres to atmospheric pressure in PSI.
 * @param {number} alt_m
 * @returns {number}
 */
function altToPsi(alt_m) {
    if (alt_m === 0) return P_SEA_LEVEL_PSI;
    const x = (alt_m / 0.3048) / 10000.0;
    return ((0.547462 * x - 7.67923) * x + 29.9309) * 0.491154 / (0.10803 * x + 1.0);
}

// Pre-computed pressure at each altitude level (descending: index 0 = sea level)
const P_LEVELS = Array.from({ length: N_ALT }, (_, i) => altToPsi(i * ALT_STEP));

// ============================================================
// CHEBYSHEV POLYNOMIAL COEFFICIENTS  (loaded once)
// ============================================================

/** Float64Array of shape [N_ALT × N_COEFF], null before init */
let _coeffs = null;

/**
 * Load the Chebyshev coefficient table from a binary file.
 * Must be awaited before calling merkelKaVL().
 *
 * @param {string} binPath  URL to merkel_poly.bin
 * @returns {Promise<void>}
 */
async function initMerkelEngine(binPath) {
    try {
        const response = await fetch(binPath);
        if (!response.ok) {
            throw new Error(`HTTP ${response.status}`);
        }
        const buf = await response.arrayBuffer();
        _coeffs = new Float64Array(buf);
    } catch (error) {
        _coeffs = null;
        console.warn(`Merkel coefficient table unavailable (${binPath}). Using fallback saturation model.`);
    }
}

// ============================================================
// SATURATED VAPOUR PRESSURE LOOKUP
// ============================================================

/**
 * Clenshaw recurrence for a Chebyshev series at normalised x ∈ [−1, 1].
 * Evaluates c[0] + c[1]·T₁(x) + … + c[n]·Tₙ(x).
 *
 * @param {number} offset  Start index in _coeffs for this altitude
 * @param {number} x       Normalised temperature ∈ [−1, 1]
 * @returns {number}
 */
function _chebEval(offset, x) {
    const x2 = 2 * x;
    let b1 = 0.0, b2 = 0.0;
    for (let k = N_COEFF - 1; k >= 1; k--) {
        const b0 = x2 * b1 - b2 + _coeffs[offset + k];
        b2 = b1;
        b1 = b0;
    }
    return x * b1 - b2 + _coeffs[offset];
}

/**
 * Return f·Pws (PSI) via polynomial evaluation + linear pressure interpolation.
 *
 * @param {number} T_F    temperature in °F
 * @param {number} P_psi  atmospheric pressure in PSI
 * @returns {number}
 */
function fPwsFromPoly(T_F, P_psi) {
    if (!_coeffs || _coeffs.length < (N_ALT * N_COEFF)) {
        const T_C = (T_F - 32) / 1.8;
        const pws_kPa = 0.61078 * Math.exp((17.2694 * T_C) / (T_C + 237.29));
        const pws_psi = pws_kPa * 0.14503773773;
        return Math.min(pws_psi * 1.0005, P_psi * 0.98);
    }

    const x = (T_F - T_MID) / T_HALF;   // normalise to [−1, 1]

    // Locate altitude bracket (P_LEVELS is descending)
    let lo;
    if (P_psi >= P_LEVELS[0]) {
        lo = 0;
    } else if (P_psi <= P_LEVELS[N_ALT - 1]) {
        lo = N_ALT - 2;
    } else {
        lo = 0;
        for (let k = 0; k < N_ALT - 1; k++) {
            if (P_psi >= P_LEVELS[k + 1]) { lo = k; break; }
        }
    }
    const hi = lo + 1;

    // Linear pressure interpolation of ln(fPws)
    const frac   = (P_psi - P_LEVELS[lo]) / (P_LEVELS[hi] - P_LEVELS[lo]);
    const ln_lo  = _chebEval(lo * N_COEFF, x);
    const ln_hi  = _chebEval(hi * N_COEFF, x);
    return Math.exp(ln_lo + frac * (ln_hi - ln_lo));
}

// ============================================================
// SATURATED AIR ENTHALPY  (replicates 0x407723)
// ============================================================

/**
 * @param {number} P_psi
 * @param {number} T_F
 * @returns {number} BTU/lb dry air
 */
function hSatImperial(P_psi, T_F) {
    const fPws = fPwsFromPoly(T_F, P_psi);
    const denom = P_psi - fPws;
    if (denom <= 0) return 999.0;
    const Ws = 0.62198 * fPws / denom;
    return CP_DRY_AIR * T_F + Ws * (H_FG + CP_STEAM * T_F);
}

// ============================================================
// MERKEL KaV/L  (replicates 0x409BFE)
// ============================================================

/**
 * Calculate the Merkel cooling tower characteristic (KaV/L).
 *
 * @param {number} hwt    Hot Water Temperature [°C]
 * @param {number} cwt    Cold Water Temperature [°C]
 * @param {number} wbt    Wet Bulb Temperature [°C]
 * @param {number} lg     Liquid-to-Gas ratio
 * @param {number} [alt_m=0]  Altitude [m]
 * @returns {{ kavl, range, approach, P_kPa, valid, error }}
 */
function merkelKaVL(hwt, cwt, wbt, lg, alt_m = 0.0) {
    if (hwt <= cwt) {
        return { kavl: 0, range: 0, approach: 0, P_kPa: 0, valid: false,
                 error: 'HWT must be greater than CWT' };
    }
    if (cwt <= wbt) {
        return { kavl: 0, range: hwt - cwt, approach: 0, P_kPa: 0, valid: false,
                 error: 'CWT must be greater than WBT' };
    }
    if (lg <= 0) {
        return { kavl: 0, range: hwt - cwt, approach: cwt - wbt, P_kPa: 0,
                 valid: false, error: 'L/G must be positive' };
    }

    const WBT_F      = wbt * 1.8 + 32;
    const Range_F    = (hwt - cwt) * 1.8;
    const Approach_F = (cwt - wbt) * 1.8;
    const CWT_F      = WBT_F + Approach_F;

    if (CWT_F + Range_F >= 212.0) {
        return { kavl: 999.0, range: hwt - cwt, approach: cwt - wbt,
                 P_kPa: 101.325, valid: false, error: 'HWT exceeds boiling point' };
    }

    const P_psi = altToPsi(alt_m);
    const P_kPa = P_psi / 0.14503773773;

    const h_in  = hSatImperial(P_psi, WBT_F);
    const h_out = Range_F * lg + h_in;

    let sum = 0;
    for (let i = 0; i < 4; i++) {
        const cheb    = CHEBYSHEV_POINTS[i];
        const h_sat_i = hSatImperial(P_psi, Range_F * cheb + CWT_F);
        const h_air_i = (h_out - h_in) * cheb + h_in;
        const df      = h_sat_i - h_air_i;
        if (df <= 0) {
            return { kavl: 999.0, range: hwt - cwt, approach: cwt - wbt,
                     P_kPa: parseFloat(P_kPa.toFixed(3)), valid: false,
                     error: 'Insufficient driving force (h_sat ≤ h_air)' };
        }
        sum += 0.25 / df;
    }

    return {
        kavl:     parseFloat((sum * Range_F).toFixed(5)),
        range:    parseFloat((hwt - cwt).toFixed(2)),
        approach: parseFloat((cwt - wbt).toFixed(2)),
        P_kPa:    parseFloat(P_kPa.toFixed(3)),
        valid:    true,
        error:    null
    };
}

export { initMerkelEngine, merkelKaVL, hSatImperial, altToPsi };

File: cti_dashboard_pro/app/backend/core/merkel_engine.py
import math
import struct
import os

P_SEA_LEVEL_PSI  = 14.696
CHEBYSHEV_POINTS = [0.9, 0.6, 0.4, 0.1]
CP_DRY_AIR       = 0.24
H_FG             = 1061.0
CP_STEAM         = 0.444

T_MID    = 124.997
T_HALF   = 74.997
N_COEFF  = 19
ALT_STEP = 10
N_ALT    = 201

def alt_to_psi(alt_m):
    if alt_m == 0: return P_SEA_LEVEL_PSI
    x = (alt_m / 0.3048) / 10000.0
    return ((0.547462 * x - 7.67923) * x + 29.9309) * 0.491154 / (0.10803 * x + 1.0)

P_LEVELS = [alt_to_psi(i * ALT_STEP) for i in range(N_ALT)]

_coeffs = None

def init_merkel_engine(bin_path):
    global _coeffs
    try:
        with open(bin_path, 'rb') as f:
            buf = f.read()
            count = len(buf) // 8
            _coeffs = struct.unpack(f"{count}d", buf)
    except Exception as e:
        _coeffs = None
        print(f"Merkel coefficient table unavailable ({bin_path}). Using fallback: {e}")

def _cheb_eval(offset, x):
    x2 = 2 * x
    b1 = 0.0
    b2 = 0.0
    for k in range(N_COEFF - 1, 0, -1):
        b0 = x2 * b1 - b2 + _coeffs[offset + k]
        b2 = b1
        b1 = b0
    return x * b1 - b2 + _coeffs[offset]

def fpws_from_poly(t_f, p_psi):
    if not _coeffs or len(_coeffs) < (N_ALT * N_COEFF):
        t_c = (t_f - 32) / 1.8
        pws_kpa = 0.61078 * math.exp((17.2694 * t_c) / (t_c + 237.29))
        pws_psi = pws_kpa * 0.14503773773
        return min(pws_psi * 1.0005, p_psi * 0.98)

    x = (t_f - T_MID) / T_HALF

    lo = 0
    if p_psi >= P_LEVELS[0]:
        lo = 0
    elif p_psi <= P_LEVELS[N_ALT - 1]:
        lo = N_ALT - 2
    else:
        for k in range(N_ALT - 1):
            if p_psi >= P_LEVELS[k + 1]:
                lo = k
                break
    hi = lo + 1

    frac = (p_psi - P_LEVELS[lo]) / (P_LEVELS[hi] - P_LEVELS[lo])
    ln_lo = _cheb_eval(lo * N_COEFF, x)
    ln_hi = _cheb_eval(hi * N_COEFF, x)
    return math.exp(ln_lo + frac * (ln_hi - ln_lo))

def h_sat_imperial(p_psi, t_f):
    fpws = fpws_from_poly(t_f, p_psi)
    denom = p_psi - fpws
    if denom <= 0: return 999.0
    ws = 0.62198 * fpws / denom
    return CP_DRY_AIR * t_f + ws * (H_FG + CP_STEAM * t_f)

def merkel_kavl(hwt, cwt, wbt, lg, alt_m=0.0):
    if hwt <= cwt:
        return {"kavl": 0, "range": 0, "approach": 0, "P_kPa": 0, "valid": False,
                "error": "HWT must be greater than CWT"}
    if cwt <= wbt:
        return {"kavl": 0, "range": hwt - cwt, "approach": 0, "P_kPa": 0, "valid": False,
                "error": "CWT must be greater than WBT"}
    if lg <= 0:
        return {"kavl": 0, "range": hwt - cwt, "approach": cwt - wbt, "P_kPa": 0,
                "valid": False, "error": "L/G must be positive"}

    wbt_f = wbt * 1.8 + 32
    range_f = (hwt - cwt) * 1.8
    approach_f = (cwt - wbt) * 1.8
    cwt_f = wbt_f + approach_f

    if cwt_f + range_f >= 212.0:
        return {"kavl": 999.0, "range": hwt - cwt, "approach": cwt - wbt,
                "P_kPa": 101.325, "valid": False, "error": "HWT exceeds boiling point"}

    p_psi = alt_to_psi(alt_m)
    p_kpa = p_psi / 0.14503773773

    h_in = h_sat_imperial(p_psi, wbt_f)
    h_out = range_f * lg + h_in

    sum_val = 0
    for i in range(4):
        cheb = CHEBYSHEV_POINTS[i]
        h_sat_i = h_sat_imperial(p_psi, range_f * cheb + cwt_f)
        h_air_i = (h_out - h_in) * cheb + h_in
        df = h_sat_i - h_air_i
        if df <= 0:
            return {"kavl": 999.0, "range": hwt - cwt, "approach": cwt - wbt,
                    "P_kPa": round(p_kpa, 3), "valid": False,
                    "error": "Insufficient driving force (h_sat <= h_air)"}
        sum_val += 0.25 / df

    return {
        "kavl": round(sum_val * range_f, 5),
        "range": round(hwt - cwt, 2),
        "approach": round(cwt - wbt, 2),
        "P_kPa": round(p_kpa, 3),
        "valid": True,
        "error": None
    }
